File: androidApp/src/main/java/com/cacd2/cacdgame/android/features/gameplay/ui/cancelgame/GamePlayCancelGameView.kt
package com.cacd2.cacdgame.android.features.gameplay.ui.cancelgame

import android.content.res.Configuration
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.wrapContentHeight
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Button
import androidx.compose.material.MaterialTheme
import androidx.compose.material.OutlinedButton
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.constraintlayout.compose.ConstraintLayout
import com.cacd2.cacdgame.AppContext
import com.cacd2.cacdgame.android.AppTheme
import com.cacd2.cacdgame.android.R
import com.cacd2.cacdgame.android.fonts
import com.cacd2.cacdgame.android.username
import org.koin.compose.koinInject

/**
 * Created by francois.dabonot@cacd2.fr on 24/05/2023.
 */
@Composable
fun GamePlayCancelGameView(
    modifier: Modifier = Modifier,
    appContext: AppContext = koinInject(),
    onCancel: () -> Unit,
    onValid: () -> Unit
) {
    Surface(modifier = modifier, color = MaterialTheme.colors.surface.copy(alpha = 0.7f)) {
        ConstraintLayout {
            val (actionContainer, image) = createRefs()
            Surface(
                modifier =
                Modifier.constrainAs(actionContainer) {
                    start.linkTo(parent.start)
                    end.linkTo(parent.end)
                    bottom.linkTo(parent.bottom)
                }.shadow(
                    elevation = 16.dp,
                    shape = RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp)
                ).wrapContentHeight()
            ) {
                Column(
                    modifier = Modifier.wrapContentHeight(),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Text(
                        text =
                        stringResource(
                            id = R.string.quit_game_message,
                            formatArgs =
                            arrayOf(
                                appContext.username.value!!
                            )
                        ),
                        modifier =
                        Modifier
                            .fillMaxWidth(0.9f)
                            .padding(top = 100.dp, bottom = 40.dp),
                        style =
                        TextStyle(
                            fontFamily = fonts,
                            fontWeight = FontWeight.SemiBold,
                            fontSize = 16.sp
                        ),
                        textAlign = TextAlign.Center
                    )
                    Row(
                        horizontalArrangement = Arrangement.SpaceEvenly,
                        modifier = Modifier.fillMaxWidth().padding(bottom = 60.dp)
                    ) {
                        OutlinedButton(onClick = onCancel, shape = RoundedCornerShape(50)) {
                            Box(
                                modifier = Modifier,
                                contentAlignment = Alignment.Center
                            ) {
                                Text(text = stringResource(R.string.button_cancel).uppercase())
                            }
                        }
                        Button(onClick = onValid, shape = RoundedCornerShape(50)) {
                            Box(
                                modifier = Modifier,
                                contentAlignment = Alignment.Center
                            ) {
                                Text(text = stringResource(R.string.button_yes).uppercase())
                            }
                        }
                    }
                }
            }
            Image(
                painter = painterResource(id = R.drawable.interruptibility),
                contentDescription = null,
                modifier =
                Modifier.constrainAs(image) {
                    bottom.linkTo(actionContainer.top, margin = (-75).dp)
                    start.linkTo(parent.start, margin = 100.dp)
                    end.linkTo(parent.end, margin = 100.dp)
                }
            )
        }
    }
}

@Preview(
    device = "id:Nexus 6",
    locale = "fr",
    uiMode = Configuration.UI_MODE_NIGHT_YES or Configuration.UI_MODE_TYPE_NORMAL
)
@Composable
fun GamePlayCancelGamePreview() {
    AppTheme {
        Surface(
            modifier = Modifier.fillMaxSize(),
            color = Color.Cyan
        ) {
            GamePlayCancelGameView(
                modifier = Modifier.fillMaxSize(),
                AppContext(
                    mUsername = "TROTINETTE"
                ),
                onValid = {
                },
                onCancel = {
                }
            )
        }
    }
}

File: iosApp/iosApp/Features/GamePlay/UI/CancelGame/GamePlayCancelGameView.swift
//
//  GamePlayCancelGameView.swift
//  CACDGAME
//
//  Created by François Dabonot on 19/07/2023.
//  Copyright © 2023 CACD2. All rights reserved.
//

import SwiftUI

struct GamePlayCancelGameView: View {

    @EnvironmentObject var appContext: AppContext
    let onCancel: () -> Void
    let onValid: () -> Void

    var body: some View {
        VStack {
            Spacer()
            VStack(alignment: .center) {
                Text(R.string.localizable.quit_game_message.localizedKeyWithFormat(appContext.username!))
                    .font(R.font.poppinsSemiBold.font(size: 16))
                    .multilineTextAlignment(.center)
                    .padding(.top, 100)
                    .padding(.bottom, 40)
                HStack {
                    Spacer()
                    SecondaryButton(R.string.localizable.button_cancel.localizedKey, action: onCancel)
                    Spacer()
                    PrimaryButton(R.string.localizable.button_yes.localizedKey, action: onValid)
                    Spacer()
                }
                .padding(.bottom, 60)
            }
            .overlay {
                R.image.cancelGame.image
                    .padding(.bottom, 330)
            }
            .background(
                Color(UIColor.systemBackground)
                    .cornerRadius(16)
                    .shadow(color: R.color.onSurface.color.opacity(0.20),
                            radius: 4)
            )
        }
        .background(R.color.surface.color.opacity(0.7))
        .edgesIgnoringSafeArea(.bottom)
    }
}

#Preview {
    VStack {
        GamePlayCancelGameView(onCancel: {}, onValid: {})
    }
    .background(.red)
    .environmentObject(AppContext(username: "DEBUG"))
}
